game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/aa/OffensiveAaFireTest.java

```diff
@@ -0,0 +1,66 @@
+package games.strategy.triplea.delegate.battle.steps.fire.aa;
+
+import static games.strategy.triplea.delegate.battle.FakeBattleState.givenBattleStateBuilder;
+import static games.strategy.triplea.delegate.battle.steps.BattleStepsTest.givenUnitWithTypeAa;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import java.util.List;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class OffensiveAaFireTest {
+
+  @Mock ExecutionStack executionStack;
+  @Mock IDelegateBridge delegateBridge;
+  @Mock BattleActions battleActions;
+
+  @ParameterizedTest(name = "[{index}] {0} is {2}")
+  @MethodSource
+  void testWhatIsValid(
+      final String displayName, final BattleState battleState, final boolean expected) {
+    final OffensiveAaFire underTest = new OffensiveAaFire(battleState, battleActions);
+    assertThat(underTest.valid(), is(expected));
+    if (expected) {
+      assertThat(underTest.getNames(), hasSize(3));
+    } else {
+      assertThat(underTest.getNames(), hasSize(0));
+    }
+  }
+
+  static List<Arguments> testWhatIsValid() {
+    return List.of(
+        Arguments.of(
+            "No Offensive Aa", givenBattleStateBuilder().offensiveAa(List.of()).build(), false),
+        Arguments.of(
+            "Some Offensive Aa",
+            givenBattleStateBuilder().offensiveAa(List.of(givenUnitWithTypeAa())).build(),
+            true));
+  }
+
+  @Test
+  void testFiringAaGuns() {
```

Comment:
nit: `test` prefix is a bit of an old convention (IIRC Junit3 required test methods to begin with 'test'). Latest convention is to focus on describing what is being verified. IE: `verifyFireOffensiveAaGunActionInvoked`. The example I think also demonstrates a question of "under which conditions should that be true?"

Should there be a second test case, or do we legitimately again assume that `isValid()` is invoked first? EG:

```
@Nested
class ExecuteOffensiveAaFire {
   @Test
    @DisplayName("Execute will not invoke fire Offensive Aa Guns when no offensive AA Exist")
   void noOffensiveAa {

   }

   @Test
    @DisplayName("Execute will invoke fire Offensive Aa Guns When Exists Offensive Aa")
   void hasOffensiveAa() {

   }
}
```

(side-note, reading the display name, and or a un-camel-cased list of the test methods and display names in a test class should practically form a complete spec for the class under test,  eg:

- not valid with no offensive Aa
- valid with offensive aa
- execute will not invoke fire Offensive Aa Guns when no offensive AA exist
- execute will invoke fire Offensive Aa Guns When Exists Offensive Aa
)